Add support for Measure and Reset in Target's C API

Cryoris opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary

Follow up on the Target C API: Add support for directives.

### Details and comments

This keeps the existing interface in place and simply extends the target entries to not only hold a `StandardGate`, but a
```rust
enum StandardOperation {
    Gate(StandardGate),
    Directive(StandardInstruction),
}
```
instead. This could potentially also be upgraded to `OperationRef` if we wanted to add support for custom gates in the target.

Docs and other constructors than measure to follow.

Cryoris committed:
add support for measure

raynelfss commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -394,30 +396,62 @@ pub unsafe extern "C" fn qk_target_free(target: *mut Target) {
         }
     }
+
+    pub fn measure() -> Self {
```
**Nitpick:** Do we want to keep the `new` convention going here? We could still call it `new_measure`.
```suggestion
    pub fn new_measure() -> Self {
```

raynelfss commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -394,30 +396,62 @@ pub unsafe extern "C" fn qk_target_free(target: *mut Target) {
+        }
+    }
 }
 
```
We should probably just have a constructor based on a `StandardInstruction` while the other should be with a `StandardGate`, perhaps more closely to this:

```rs
pub fn new_instruction(instruction: StandardInstruction) -> TargetEntry {
    Self {
        operation: StandardOperation::Instruction(instruction),
        params: Some(params),
        map: Default::default(),
    }
}
```
We can then have the different variants for measure, reset, and delay as functions exposed through the FFI.
```rust
#[no_mangle]
#[cfg(feature = "cbinding")]
pub extern "C" fn qk_target_entry_measure() -> *mut TargetEntry {
    Box::into_raw(Box::new(TargetEntry::new_instruction(StandardInstruction::Measure)))
}

#[no_mangle]
#[cfg(feature = "cbinding")]
pub extern "C" fn qk_target_entry_reset() -> *mut TargetEntry {
    Box::into_raw(Box::new(TargetEntry::new_instruction(StandardInstruction::Reset)))
}

#[no_mangle]
#[cfg(feature = "cbinding")]
pub extern "C" fn qk_target_entry_delay() -> *mut TargetEntry {
    Box::into_raw(Box::new(TargetEntry::new_instruction(StandardInstruction::Delay)))
}
```

Cryoris committed:
add delay/reset

raynelfss committed:
Refactor: `TargetEntry` constructor from gates or instructions.
- Remove extra `TargetEntry` rust initializers in favor of one `TargetEntry::new_instruction`.
- Since most instructions don't require parameters, we do not need to add an option for fixed parameters.

raynelfss committed:
Remove: `Delay` support in Target C FFI.
- After internal discussion we decided to not pursue adding `Delay` support in this PR due to its reliance in `ParameterExpression`. It will be re-added in a future update.
- Rename special `qk_target_entry` functions to have `new` as part of the suffix.

raynelfss committed:
Fix: Incorrect extraction of `acquire_alignment` in Target creation

raynelfss committed:
Fix: Failing updated tests forget to compare with the correct number.

mtreinish commented on crates/cext/src/circuit.rs:
```diff
@@ -1043,13 +1055,7 @@ pub unsafe extern "C" fn qk_circuit_delay(
-        QkDelayUnit::NS => DelayUnit::NS,
-        QkDelayUnit::PS => DelayUnit::PS,
-    };
+    let delay_unit_variant = unit.into();
```
Given the release deadline on this I'm fine with leaving this as is, but it is an unrelated change and shouldn't be in this PR.

mtreinish commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -444,6 +477,38 @@ pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEnt
+///
+/// # Example
+///
+///     QkTargetEntry *entry = qk_target_entry_new_measure()();
```
```suggestion
///     QkTargetEntry *entry = qk_target_entry_new_measure();
```

mtreinish commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -444,6 +477,38 @@ pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEnt
+///
+/// # Example
+///
+///     QkTargetEntry *entry = qk_target_entry_new_reset();
```
These examples aren't much of an example. Maybe show adding a couple qubits with an error rate and adding that to a target?

raynelfss replied:
Added in 0640bbc

mtreinish commented on crates/transpiler/src/target/mod.rs:
```diff
@@ -305,7 +305,7 @@ impl Target {
             min_length: min_length.unwrap_or(1),
             pulse_alignment: pulse_alignment.unwrap_or(1),
-            acquire_alignment: acquire_alignment.unwrap_or(0),
+            acquire_alignment: acquire_alignment.unwrap_or(1),
```
This is also an unrelated fix right? It's fine given the timeframe but it would be better as a standalone PR.

raynelfss replied:
Yeah, for this one it was a bug I caught just now which would've surfaced after #14308 merged. There was some incorrect initialization happening in the rust-side target.

mtreinish commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -444,6 +477,38 @@ pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEnt
 }
 
+/// @ingroup QkTargetEntry
+/// Creates a measurement entry for a ``QkTarget``.
```
```suggestion
/// Creates a new entry for adding a measurement instruction to a ``QkTarget``.
```

mtreinish commented on crates/cext/src/transpiler/target.rs:
```diff
@@ -444,6 +477,38 @@ pub extern "C" fn qk_target_entry_new(operation: StandardGate) -> *mut TargetEnt
+}
+
+/// @ingroup QkTargetEntry
+/// Creates a reset entry for a ``QkTarget``.
```
```suggestion
/// Creates a new entry for adding a reset instruction to a ``QkTarget``.
```

mtreinish reviewed (commented):
Overall this LGTM code-wise. I think the docs need a few updates though.

raynelfss committed:
Apply suggestions from code review

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Docs: Add more throrough examples to docstrings.

mtreinish reviewed (approved):
LGTM, thanks for the updates.
